common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Extensions/DataProtectionBuilderEx.cs

```diff
@@ -0,0 +1,218 @@
+                    return keyVault;
+                }
+            }
+
+            // Try using aims
+            keyVault = TryInititalizeKeyAsync("Managed Service Identity",
+                vaultUri, keyName).Result;
+            if (keyVault != null) {
+                return keyVault;
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Read configuration secret
+        /// </summary>
+        /// <param name="method"></param>
+        /// <param name="vaultUri"></param>
+        /// <param name="keyName"></param>
+        /// <param name="connectionString"></param>
+        /// <returns></returns>
+        private static async Task<KeyVaultClient> TryInititalizeKeyAsync(
+            string method, string vaultUri, string keyName, string connectionString = null) {
+            var tokenProvider = new AzureServiceTokenProvider(connectionString);
+            try {
+                var keyVaultClient = new KeyVaultClient(
+                    new KeyVaultClient.AuthenticationCallback(
+                        tokenProvider.KeyVaultTokenCallback));
+
+                try {
+                    var key = await keyVaultClient.GetKeyAsync(vaultUri, keyName);
+                }
+                catch {
+                    // Try create key
+                    await keyVaultClient.CreateKeyAsync(vaultUri, keyName, new NewKeyParameters {
+                        KeySize = 2048,
+                        Kty = JsonWebKeyType.Rsa,
+                        KeyOps = new List<string> {
+                            JsonWebKeyOperation.Wrap, JsonWebKeyOperation.Unwrap
+                        }
+                    });
+                }
+                // Worked - we have a working keyvault client.
+                return keyVaultClient;
+            }
+            catch (Exception ex) {
+                Log.Logger.Debug("Failed to authenticate to keyvault {url} using " +
+                    "{method}: {message}",
+                    vaultUri, method, ex.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Dataprotection default configuration
+        /// </summary>
+        internal sealed class DataProtectionConfig : ConfigBase, IClientConfig,
+            IKeyVaultConfig, IStorageConfig {
+
+            /// <summary>Application id</summary>
+            public string AppId => GetStringOrDefault("PCS_KEYVAULT_APPID",
+                Environment.GetEnvironmentVariable("PCS_KEYVAULT_APPID"))?.Trim();
+            /// <summary>App secret</summary>
+            public string AppSecret => GetStringOrDefault("PCS_KEYVAULT_SECRET",
+                Environment.GetEnvironmentVariable("PCS_KEYVAULT_SECRET"))?.Trim();
+            /// <summary>Optional tenant</summary>
+            public string TenantId => GetStringOrDefault("PCS_AUTH_TENANT",
+                Environment.GetEnvironmentVariable("PCS_AUTH_TENANT") ?? "common").Trim();
```

Comment:
string dupes, better make this a constant somewhere.

Reply:
Done